src/component/task_list.rs

```diff
@@ -151,30 +151,79 @@ impl DrawableComponent for TaskList {
                 if app.task_store.tasks.is_empty() {
                     return EventResult::Ignored;
                 }
-                app.task_store.tasks[*selected_index].priority = app.task_store.tasks
-                    [*selected_index]
-                    .priority
-                    .next_priority();
+
+                let old_task = {
+                    let task = &mut app.task_store.tasks[*selected_index];
+
+                    task.priority = task.priority.next_priority();
+
+                    task.clone()
+                };
+
+                if app.task_store.auto_sort {
+                    app.task_store.sort();
+                }
+
+                *selected_index = app
+                    .task_store
+                    .tasks
+                    .iter()
+                    .position(|t| *t == old_task)
+                    .expect("getting task index after sorting")
+                    .to_owned();
             }
             KeyCode::Char('J') => {
-                let task_length = app.task_store.tasks.len();
-                let task = app.task_store.tasks.remove(*selected_index);
-                app.task_store
-                    .tasks
-                    .insert((*selected_index + 1) % task_length, task);
-                *selected_index = (*selected_index + 1) % task_length;
+                if app.task_store.auto_sort {
+                    let tasks_length = app.task_store.tasks.len();
+                    let task = &app.task_store.tasks[*selected_index];
+                    let task_below = &app.task_store.tasks[*selected_index + 1];
+
+                    if task.priority == task_below.priority {
+                        let task = app.task_store.tasks.remove(*selected_index);
+
+                        let new_index = (*selected_index + 1) % tasks_length;
+
+                        app.task_store.tasks.insert(new_index, task);
+
+                        *selected_index = new_index;
+                    }
+                } else {
+                    let task = app.task_store.tasks.remove(*selected_index);
+                    let tasks_length = app.task_store.tasks.len();
+                    let new_index = (*selected_index + 1) % tasks_length;
+
+                    app.task_store.tasks.insert(new_index, task);
+
+                    *selected_index = new_index;
+                }
             }
             KeyCode::Char('K') => {
-                let task_length = app.task_store.tasks.len();
-                let task = app.task_store.tasks.remove(*selected_index);
-                if *selected_index == 0 {
-                    app.task_store.tasks.insert(task_length - 1, task);
-                    *selected_index = task_length - 1;
+                if app.task_store.auto_sort {
```

Comment:
Just one last thing, this and the one above can be refactored really nicely (above doesn't contain conversion to isize/adds one instead), it also resolves a crash when `auto_sort` was disabled.
```rust
            KeyCode::Char('K') => {
                let tasks_length = app.task_store.tasks.len();
                let new_index = (*selected_index as isize - 1).rem_euclid(tasks_length as isize) as usize;

                let task = &app.task_store.tasks[*selected_index];
                let task_above = &app.task_store.tasks[new_index];

                if task.priority == task_above.priority || !app.task_store.auto_sort {
                    let task = app.task_store.tasks.remove(*selected_index);
                    app.task_store.tasks.insert(new_index, task);
                    *selected_index = new_index;
                }
            }
```

Reply:
Wow, that's smart! I've seen `rem_euclid()` before, but I wrote it off as "just another math-y thing". That's actually really handy.
And yeah, I goofed up with the overflow check there... Thanks for a great suggestion. Wow!)

Reply:
Yea, I recently saw it on a YT Short, and was like hang on I could use this :)
There seems to be an edge case where it would crash if there are no tasks that I didn't consider :(
Really sorry about that

Reply:
No, I still think that this looks much better and we could just add a
```rust
if tasks_length == 0 { return EventResult::Ignored }
````
 right?

Reply:
Yep, should be good!

Reply:
The `len() == 0` thing is also the case with the 'J' one too, so that's actually my oversight :)